### ISSUE TITLE
NamedScratchpads somewhat broken.

### ISSUE THREAD

**Issue by exorcist365** [xmonad/xmonad-contrib#728]
### Problem Description

After commit 3fc830aa09368dca04df24bf7ec4ac817f2de479 my scratchpads suddenly stopped working after I restarted the window manager, it does however fix itself after opening a new window but having to do that after every time I restart the wm and want to bring my scratchpad is kinda... annoying, don't you think?

### Steps to Reproduce

Run a version of xmonad-contrib pushed after the 1st of April, restart the window manager and then press your scratchpad keybind.

### Configuration File

Quite difficult to post here since my setup is split into multiple files, so you better take a look at my [dotfiles](https://gitlab.com/exorcist365/dotfulls/-/tree/master/.config/xmonad/src)

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - I tested my configuration
    - [x] With `xmonad` version 0.17 (commit master)
    - [x] With `xmonad-contrib` version 0.17 (commit master and the one mentioned above)

**Comment by geekosaur** [comment]
Are you using NamedScratchpad or standard Scratchpad? I suspected the new scratchpad implementation would break the latter, and I'm surprised it's taken this long for a report to come in.

**Comment by geekosaur** [comment]
Sorry, I missed the title. I've been using several NamedScratchpads without issue since that commit, so I'm not sure what could be wrong.

**Comment by slotThe** [comment]
> ### Configuration File
>
> Quite difficult to post here since my setup is split into multiple files, so you better take a look at my [dotfiles](https://gitlab.com/exorcist365/dotfulls/-/tree/master/.config/xmonad/src)

Could you perhaps try to produce a (minimal) example with which you can reproduce this?  Much like @geekosaur, I've been using scratchpads extensively since this commit and I've never noticed this

**Comment by exorcist365** [comment]
ok, so:
```hs
fmSP, ghciSP, ocamlSP, musicSP, terminalSP :: String
fmSP = "FFF"
ghciSP = "GHCI"
ocamlSP = "OCaml"
musicSP = "Music"
terminalSP = "Terminal"

scratchpads :: [NamedScratchpad]
scratchpads = [ NS { name = ghciSP
                   , query = isGHCI
                   , cmd = spawnGHCI
                   , hook = centerFloat
                   }
              , NS { name = ocamlSP
                   , query = isOcaml
                   , cmd = spawnOcaml
                   , hook = centerFloat
                   }
              , NS { name = musicSP
                   , query = isMusic
                   , cmd = spawnMusic
                   , hook = centerFloat
                   }
              , NS { hook = topFloat
                   , name = terminalSP
                   , query = isTerminal
                   , cmd = spawnTerminal
                   }
              , NS { name = fmSP
                   , query = isFM
                   , cmd = spawnFM
                   , hook = centerFloat
                   }
              ] where
                  spawnFM, spawnGHCI, spawnOcaml, spawnMusic, spawnTerminal :: String
                  spawnFM = unwords [term, "-c", fmSP, "-e", files]
                  spawnGHCI = unwords [term, "-c", ghciSP, "-e", "stack exec -- ghci", "-v0", "-ghci-script", configHome </> "ghci/ghci.conf"]
                  spawnOcaml = unwords [term, "-c", ocamlSP, "-e", "myutop"]
                  spawnMusic = unwords [term, "-c", musicSP, "-e", music]
                  spawnTerminal = unwords [term, "-c", terminalSP]

                  isFM, isGHCI, isOcaml, isMusic, isTerminal :: Query Bool
                  isFM = className =? fmSP
                  isGHCI = className =? ghciSP
                  isOcaml = className =? ocamlSP
                  isMusic = className =? musicSP
                  isTerminal = className =? terminalSP

                  topFloat, centerFloat :: ManageHook
                  topFloat = customFloating $ RationalRect 0 0 1 0.45
                  centerFloat = customFloating $ RationalRect (1/6) (1/6) (2/3) (2/3)
```
thats the scratchpad part and the calling i do with the function `namedScratchpadAction`

**Comment by geekosaur** [comment]
This isn't quite enough information. You might want to compare to how I use them in [my config](https://github.com/geekosaur/xmonad.hs/blob/skkukuk/xmonad.hs); note in particular lines 56–74 (definition), 122 (`manageHook`), 155–158 (keybindings). It does sound like you might be missing the `manageHook`, although in that case I'd expect it not to work at all.

**Comment by exorcist365** [comment]
```hs
windowManager = mconcat [ tileBelow
                        , transience'
                        , manageSpawn
                        , manageHook def
                        , isRole =? "About" --> doCenterFloat
                        , namedScratchpadManageHook scratchpads
                        , isRole =? "pop-up" --> forceCenterFloat
                        , className =? "Sxiv" --> hasBorder False
                        , className =? "Picker" --> doCenterFloat
                        , isName =? "Library" --> forceCenterFloat
                        , className =? "Places" --> forceCenterFloat
                        , className =? "New VM" --> forceCenterFloat
                        , className =? "Ripcord" --> shiftFocus (at 5)
                        , className =? "discord" --> shiftFocus (at 5)
                        , className =? "Filezilla" --> shiftFocus (at 6)
                        , className =? "qutebrowser" --> shiftFocus (at 2)
                        , className =? "Signal Beta" --> shiftFocus (at 5)
                        , isDialog --> mconcat [doF siftUp, doCenterFloat]
                        , isName =? "Picture-in-Picture" --> forceCenterFloat
                        , className =? "TelegramDesktop" --> shiftFocus (at 5)
                        , className =? "mpv" --> mconcat [doF siftUp, unFloat]
                        , isRole =? "GtkFileChooserDialog" --> forceCenterFloat
                        , className =? "st-float" --> mconcat [doFocus, doFloat]
                        , className =? "Ripcord" <&&> isName =? "Preferences" --> forceCenterFloat
                        , className =? "help" --> mconcat [forceCenterFloat, doF siftUp, hasBorder False]
                        , isInProperty "_NET_WM_WINDOW_TYPE" "_NET_WM_WINDOW_TYPE_SPLASH" --> doCenterFloat
                        ]
``` this is my managehook

**Comment by slotThe** [comment]
Could you perhaps post a _complete_ minimal example that we can just
copy-paste into an xmonad.hs file?  Just snippets won't really do us any
good (as in, not much more than you posting your entire configuration).

**Comment by exorcist365** [comment]
is that good?

**Comment by geekosaur** [comment]
The root window (singular) is always mapped and behaves as if it were OverrideRedirect. You may have had some other windows waiting to be mapped.

**Comment by exorcist365** [comment]
Tho it might be just mine setup thats breaking.

**Comment by exorcist365** [comment]
Fixed by #730.
